recipes/mingw-w64/all/conanfile.py

```diff
@@ -13,29 +13,193 @@ class MingwConan(ConanFile):
     settings = "os", "arch", "compiler"
     options = {"threads": ["posix", "win32"], "exception": ["seh", "sjlj"]}
     default_options = {"threads": "posix", "exception": "seh"}
-    build_requires = "7zip/19.00"
     no_copy_source = True
 
     def validate(self):
-        if self.settings.os != "Windows":
-            raise ConanInvalidConfiguration("MinGW is only supported by Windows.")
+        valid_os = self.conan_data["sources"][self.version]["url"].keys()
+        if str(self.settings.os) not in valid_os:
+            raise ConanInvalidConfiguration(f"MinGW {self.version} is only supported for the following operating systems: {valid_os}")
+        valid_arch = self.conan_data["sources"][self.version]["url"][str(self.settings.os)].keys()
+        if str(self.settings.arch) not in valid_arch:
+            raise ConanInvalidConfiguration(f"MinGW {self.version} is only supported for "
+                                            f"the following architectures on {str(self.settings.os)}: {valid_arch}")
+
+    def configure(self):
+        if self.settings.os == "Windows":
+            self.build_requires("7zip/19.00")
 
     def source(self):
-        url = self.conan_data["sources"][self.version]["url"][str(self.settings.arch)] \
-                             [str(self.options.threads)][str(self.options.exception)]
-        self.output.info("Downloading: %s" % url["url"])
-        tools.download(url["url"], "file.7z", sha256=url["sha256"])
-        self.run("7z x file.7z")
-        os.remove('file.7z')
+        arch_data = self.conan_data["sources"][self.version]["url"][str(self.settings.os)][str(self.settings.arch)]
+
+        if self.settings.os == "Windows":
+            url = arch_data[str(self.options.threads)][str(self.options.exception)]
+            self.output.info("Downloading: %s" % url["url"])
+            tools.download(url["url"], "file.7z", sha256=url["sha256"])
+            self.run("7z x file.7z")
+            os.remove('file.7z')
+        else:
+            for package in arch_data:
+                self.output.info(f"Downloading {package} from {arch_data[package]['url']}")
+                tools.get(**arch_data[package], strip_root=True, destination=os.path.join(self.source_folder, package))
```

Comment:
code in the `source` method cannot depend on any settings/options.

Reply:
Hmm, do you have a suggestion how to fix this?

The source method is completely different between the windows and the linux OS...

Reply:
When we download binaries, that is done in `build`.
See e.g. my `arm-toolchain` pr at https://github.com/conan-io/conan-center-index/pull/6822

I'm currently thinking that this recipe is trying to do/package too much: building from sources + packaging binaries.
Would it be possible to build everything from source and ignore the pre-built binaries?

The main problem is MSVC, as that cannot compile gcc. At least, I'm not aware it is able to.
Alternatively, it might be possible to do (pseudo-code):
```python
if self.settings.compiler == "Visual Studio":
    self.build_requires("tcc")  # https://bootstrapping.miraheze.org/wiki/C_compilers#Can_tcc_build_gcc_4.7.4?_[yes]
    build_gcc_4_7_4()
    toolchain = {
         "CC": "gcc-4.7",
         "CFLAGS": "",
    }
else:
    toolchain = {
        "CC": tools.get_env("CC", str(self.settings.compiler)),
        "CFLAGS": "",
    }
# build mingw_w64 with toolchain
```

Reply:
Ok, so you would suggest to move the download also to the build step. Will change it.
I thought that the source step should be used to get the source, and then the build step to build from that source... Or did I understood that wrong?

Regarding your second comment, unfortunately I did not fully understand it.
At least for the Linux version, it is just downloading sources and building from them. No pre-built binaries.
Regarding the pre-built binaries you are probably referring to the Windows version, right? I did not want to change that one, but only add linux. Also since I cannot really test it locally on Windows.

Reply:
I'm thinking that a self-built mingw toolchain deserves its own recipe, not shared with prebuilt ones.
Perhaps this belongs in another subdirectory of `mingw-w64`, and versioned somehow.

About my windows suggestion. Let's just ignore that, because it's not something trivial that will work at the first try and can be debugged from c3i.

Reply:
The goal of the source method is that one can call source once (whatever operating system), and run `build` multiple times.
One should even be able to transport the source folder and get started building.

Reply:
Ok, thanks for the clarification. But then it still should go into the source method, the download step?
Since it is just downloading the tars, and extracting them.
Building can be called multiple times in the same sources folder. I actually used the `-k` switch to avoid downloading the tars all the time. So I do not fully understand why the tar downloading should be done in the build step?

Reply:
Note that the Autotools is creating all files in the build folder, and (as far as I understood) not changing the source folder in the current way I am using it.

Reply:
(About sources and where to execute those lines) In future Conan (v2) `self.options` and `self.settings` shouldn't be available in `def source()` method. Why? The files downloaded to the `source` folder are being used by different configurations (if you change an option value, Conan won't execute `source()` again if the files are already there). That's the reason why, if sources depend on the configuration, they should be downloaded in the `build()` method, which will use a different folder for each configuration.

Reply:
Cool, thanks for the additional explanation. Changed it in https://github.com/conan-io/conan-center-index/pull/6849/commits/2b1777f295382d32fc739302c5e9ce71292f29af